### DISCUSSION TITLE
About Special keys

### DISCUSSION THREAD

**Discussion by Mananangal** [autokey/autokey#674] [Q&A]
I am planning on using autokey to create a macro for a game. The macro requires RSHIFT+DEL+R but I couldn't find the key for rightshift on the wiki. Is there a key for it or will it be added?

**Comment by Elliria** [comment]
Keys that aren't listed in the AutoKey API can still be pressed and released as long as you know their X keycodes. In Linux, you can run the **xev** command in a terminal window and then press a key to see its data. There will be a lot of output. You're looking for the number after **keycode** in the output.

_Note that there will be duplicate data each time you press and release a key, because the xev command reacts to a key press and a key release as separate events._

To display all output from any input device:
```bash
xev
```

To display all output from only the keyboard:
```bash
xev -event keyboard
```

To display the word "keycode" followed by the keycode:
```bash
xev -event keyboard | grep -Eo ".{,0}keycode.{,4}"
```

To display only the keycode:
```bash
xev -event keyboard | grep -Po '(?<=keycode\s)[^\s]*'
```

Once you have a key's keycode, you can use AutoKey's ```press_key``` and ```release_key``` API calls. For example, to press and release the right Shift key, you would use these two calls: 
```python
keyboard.press_key("<code62>")
keyboard.release_key("<code62>")
```

For your particular need of right shift plus delete plus r, I'm not sure which of these three you would need, so you might want to try them one by one:

This example presses each key, one after another and then releases each key, one after another:
```python
keyboard.press_key("<code62>")
keyboard.press_key("<delete>")
keyboard.press_key("r")
keyboard.release_key("r")
keyboard.release_key("<delete>")
keyboard.release_key("<code62>")
```

This example presses the right shift key and then presses and releases the delete key followed by pressing and releasing the r key followed by releasing the right shift key:
```python
keyboard.press_key("<code62>")
keyboard.press_key("<delete>")
keyboard.release_key("<delete>")
keyboard.press_key("r")
keyboard.release_key("r")
keyboard.release_key("<code62>")
```
This example presses and releases the shift key followed by pressing and releasing the delete key followed by pressing and releasing the r key:
```python
keyboard.press_key("<code62>")
keyboard.release_key("<code62>")
keyboard.press_key("<delete>")
keyboard.release_key("<delete>")
keyboard.press_key("r")
keyboard.release_key("r")
```

**Reply by josephj11** [reply]
I haven't seen those xev commands before. It would be great to add this to the wiki.

There has been some discussion on enabling AutoKey to distinguish between the various "left" and "right" keys. Apparently, it's an easy enhancement. I do not see an open enhancement issue for this. Apparently, this has not been pursued.

#76 is tangentially related.

**Reply by popeliao** [reply]
Is there a way to use "right shift" to trigger an autokey event?
For instance, I want to use "right shift" + asdw to become left, down, right, up.
But I still want "left shift" plus those keep for uppercase characters.

**Reply by josephj11** [reply]
Not at the moment. We need someone to file an enhancement request for this so the devs remember to implement it. As stated above, it's probably an easy enhancement to add.

It can't make it into 0.96.0, but it could come shortly after that.

The main problem is the shortage of developer resources. We have one active developer who will become significantly less active shortly because of IRL demands on his time and a couple of others who occasionally do very good work, but that's it. If you know any Python developers, send them our way!

If you want to, you can experiment with things like xkb and xmodmap to remap one or more keys outside of AutoKey, but AutoKey isn't very partial to remapped keyboards, so it may require some trial and error.

**Comment by Elliria** [comment]
Adding the left and right versions of the keys would be a useful enhancement. This would be an addition rather than a replacement, right?

Those xev commands are a bit messy because of the excess data. I looked and looked, but wasn't able to find any examples of a command that spits out just one keycode in response to one key-press. Hopefully someone will come along with a more succinct one. Meanwhile, I'll add the above information to the [Special Keys](https://github.com/autokey/autokey/wiki/Special-Keys) wiki page.

**Comment by josephj11** [comment]
AFAIK, separate modifier keys was never filed as an enhancement issue. I am pretty sure someone looked into it though - maybe @sebastiansam55 ?

**Comment by Elliria** [comment]
Ah, okay. Thanks for the update.
